Pegasus/include/geometry.hpp

```diff
@@ -0,0 +1,880 @@
+            projectAllVertices(axis, cache->aBoxVertices.begin(), cache->aBoxVertices.end(), cache->aBoxVerticesDistances.begin());
+            projectAllVertices(axis, cache->bBoxVertices.begin(), cache->bBoxVertices.end(), cache->bBoxVerticesDistances.begin());
+            std::sort(cache->aBoxVerticesDistances.begin(), cache->aBoxVerticesDistances.end());
+            std::sort(cache->bBoxVerticesDistances.begin(), cache->bBoxVerticesDistances.end());
+
+            if (cache->aBoxVerticesDistances.back() < cache->bBoxVerticesDistances.back()) {
+                if (cache->aBoxVerticesDistances.back() < cache->bBoxVerticesDistances.front()) {
+                    return false;
+                }
+            }
+            else {
+                if (cache->bBoxVerticesDistances.back() < cache->aBoxVerticesDistances.front()) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    template <>
+    inline Vector3 calculateContactNormal<Box, Box>(SimpleShape const *a, SimpleShape const *b, CacheBase *cacheBase)
+    {
+        auto cache = static_cast<Cache<Box, Box>*>(cacheBase);
+
+        std::array<double, 6> distances;
+        for (uint32_t i = 0; i < distances.size(); ++i) {
+            distances[i] = (cache->aMassCenter - cache->bBoxFaces[i]).magnitude();
+        }
+
+        auto const minIt = std::min_element(distances.begin(), distances.end());
+        auto const minIndex = std::distance(distances.begin(), minIt);
+        cache->contactNormal = cache->bBoxAxes[minIndex].unit();
+
+        return cache->contactNormal;
+    }
+
+    template <>
+    inline double calculatePenetration<Box, Box>(SimpleShape const *a, SimpleShape const *b, CacheBase *cacheBase)
+    {       
+        auto cache = static_cast<Cache<Box, Box>*>(cacheBase);      
+
+        projectAllVertices(cache->contactNormal, cache->aBoxVertices.begin(),
+                           cache->aBoxVertices.end(), cache->aBoxVerticesDistances.begin());
+        projectAllVertices(cache->contactNormal, cache->bBoxVertices.begin(),
+                           cache->bBoxVertices.end(), cache->bBoxVerticesDistances.begin());
+
+        double bMaxVertexDistance = *std::max_element(cache->bBoxVerticesDistances.begin(), cache->bBoxVerticesDistances.end());
+        double aMinVertexDistance = *std::min_element(cache->aBoxVerticesDistances.begin(), cache->aBoxVerticesDistances.end());
+        cache->penetration = bMaxVertexDistance - aMinVertexDistance;
+
+        return cache->penetration;
+    }
+
+} // namespace intersection
+
+    // General intersection
+    using ShapeTypePair = std::pair<SimpleShapeType, SimpleShapeType>;
+
+    size_t shapeTypePairHash(ShapeTypePair const & p);
+
+    class IntersectionQuery {
+    private:
+        std::unordered_map<ShapeTypePair,
+            std::unique_ptr<intersection::CacheBase>,
+            std::function<size_t(ShapeTypePair const & p)> >
```

Comment:
Why use `std::function` where we don't need to erase type information? A simple `struct` with `operator()` would work just as fine if not better